gcc: use standard builder instead of custom builder.sh

amjoseph-nixpkgs opened the pull request:
**Please review the commits in this PR separately, ignoring whitespace**.

If you read them that way, most of them are trivial.  You can skip the first six commits, which are from this PR (merged to `master` but has not yet reached `staging`):

- https://github.com/NixOS/nixpkgs/pull/238154

##### Description of changes

This PR causes our gcc expressions (all 10 of them) to use the standard builder instead of their own custom `builder.sh`.

This PR deliberately does the minimum amount possible to achieve that, because it is likely to merge-conflict with almost any other PR which touches any version of GCC.  Most of the commits in this PR are simple, repeatable operations (like search and replace) to try to make it easier to rebase them past any conflicts.

Most of the benefits of this PR will be realized in future PRs (see the last few commits of https://github.com/NixOS/nixpkgs/pull/238069)


##### Future Benefits

- Right now changing even a single bit in `gcc/builder.sh` causes a mass rebuild of every package on every platform.  This is crazy.
  - I would like to do a significant amount of work on our cross-compilation scheme, and almost none of that work should trigger mass rebuilds or have to go through `staging`.
- Overriding build phases is painful with `builder.sh`; see the [gymnastics with "extra fixup phases"](https://github.com/NixOS/nixpkgs/blob/d403dd77cedc41b31bfe99564c08fdcfddcc05aa/pkgs/development/compilers/gcc/common/libgcc.nix#L17-L19) I had to add to `libgcc.nix`.  Those will go away.
- We're currently passing a ton of variables into the builder environment, many of which could be accessed directly by Nix conditionals.

And, ultimately the big long-term benefit:

- Stop maintaining 10 copies of everything, and risking subtle hard-to-notice bugs if one copy gets out of sync with the others.  Our `gcc` expression should look like our [`boost`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/boost/generic.nix) and [`nix`](https://github.com/NixOS/nixpkgs/blob/a056d3567cfa95b732d13a34347ae8babc3deb94/pkgs/tools/package-management/nix/common.nix#L164-L187) expressions, using `lib.versions` when it is necessary to set minimum versions for things like patch applicability or configure flags.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux `gcc48,gcc49,gcc6,gcc7,gcc8,gcc9,gcc10,gcc11,gcc12,gcc13`
  - [ ] aarch64-linux (cross from `x86_64-linux`): in progress
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
gcc: copy gcc/builder.sh to gcc/common/builder.nix

This commit is left unsquashed to make review of the subsequent
commits easier.

Commit:
gcc: make common/builder.nix into valid Nix syntax

This commit performs two search-and-replace operations:

- replace all `${` with `''${`
- replace `''` with `""` in shell scripts

This commit is left unsquashed to make review of the subsequent
commits easier.

Commit:
gcc: add args to common/builder.nix, split up script into phases

This commit adds arguments to `builder.nix`, making it a callable
function, and splits up the single massive shell script into
separate attributes for each phase.  It also drops the first two
lines and the last line because these are part of the default
builder.

All script lines which were not part of a phase function have been
moved into `preUnpack` since this is the first phase that runs.
Subsequent commits will move parts of this to more sensible
locations.

Commit:
gcc: remove `builder = ../builder.sh` from gcc/${version}/default.nix

This commit deletes the custom builder attribute from all gcc versions.

Commit:
gcc: use common/builder.nix

This commit replaces `stdenv.mkDerivation` with `(callPackage
../common/builder.nix {})` in all versions of gcc.

amjoseph-nixpkgs commented:
@ofborg build

Commit:
gcc: remove builder.sh

Commit:
gcc: add `-g` to `declare`

The bash `declare` builtin works differently when it occurs inside
function, as it now does due to breaking up the monolithic
`builder.sh` into separate phases.  We have to add `-g` in order to
get the outside-of-a-bash-function behavior when using `declare`
within a bash function.

trofi reviewed (dismissed):
Looks good.

`diffoscope` does not show any material changes `before`/`after` at least for `gcc.cc`.

trofi commented:
Having said that `wineWowPackages.waylandFull.stdenv.cc.cc` (a dependency of `wineWowPackages.waylandFull`) somehoow fails to build for me now. Has something to do with library search path (might also be my local patches):

```
/nix/store/163svia93c95033jqv72yin4vwm4s8v8-binutils-patchelfed-ld-2.40/bin/ld: skipping incompatible /nix/store/8gf90snmvzp1sg5h68lrn1qiw5vchnfg-glibc-2.37-8/lib/libc.so when searching for -lc
```

amjoseph-nixpkgs commented:
> Having said that `wineWowPackages.waylandFull.stdenv.cc.cc` (a dependency of `wineWowPackages.waylandFull`) somehoow fails to build for me now.

Yeah, I'm seeing this too.  Investigating.

Commit:
gcc: use Nix instead of bash for conditional

Now that we use the standard builder, the commands produced by
pre-configure.nix are wrapped in a bash function.  Inside of a bash
function, `export foo=` will still add `foo` to the environment of
any child processes forked after that point, but those variables
will *not* be visible to bash code which is outside of the
function-scope in which the `export` occurs.

Weird crap like this is yet another reason why we need to move away
from using bash for logic.  Let's switch.

Commit:
gcc: use callFile instead of callPackage for builder.nix

trofi commented:
Looks like a bunch of unrelated commits were pushed to this branch.

amjoseph-nixpkgs commented:
> Looks like a bunch of unrelated commits were pushed to this branch.

Yeah this needs to wait until #238154 is merged; trying to rebase them past each other is way too error-prone.

amjoseph-nixpkgs commented:
Rebased.

amjoseph-nixpkgs commented:
> Having said that `wineWowPackages.waylandFull.stdenv.cc.cc` (a dependency of `wineWowPackages.waylandFull`) somehoow fails to build for me now. Has something to do with library search path (might also be my local patches):

Since rebasing this I am now able to build that package.

```
$ nix-instantiate . -A wineWowPackages.waylandFull.stdenv.cc.cc
/nix/store/z4pmqv96qlvq582vkcgyq62mqz64ajkn-gcc-12.3.0.drv
$ nix build -f . wineWowPackages.waylandFull.stdenv.cc.cc
$ git rev-parse HEAD
6cdf114d63d500bbdb7c43f6e2cbc4149e1b90ea
```

@ofborg build wineWowPackages.waylandFull.stdenv.cc.cc

amjoseph-nixpkgs commented:
Okay, I found the problem with `gcc_multi`, so I think this PR is ready.

Since last review I have:
1. Rebased (no other changes)
2. Pushed c76e34af5cbc6364d11e64c053a9b751d901f51d which simply replaces `callPackage` with `callFile`
3. Pushed e02be1883fb2f1eacde8b193400c2168fce20737 which fixes the problem with `wineWowPackages.waylandFull.stdenv.cc.cc`.

So I think e02be1883fb2f1eacde8b193400c2168fce20737 is the only commit that needs additional review.

```
$ git rev-parse HEAD
e02be1883fb2f1eacde8b193400c2168fce20737
$ nix-instantiate . -A wineWowPackages.waylandFull.stdenv.cc.cc
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
/nix/store/3i9dfhm3kwn269gblv0clqh05lypflm6-gcc-12.3.0.drv
$ nix build -f . wineWowPackages.waylandFull.stdenv.cc.cc && echo ok
ok
```

@ofborg build wineWowPackages.waylandFull.stdenv.cc.cc

amjoseph-nixpkgs commented:
Building:

- [ ] `gcc48 gcc49 gcc6 gcc7 gcc8 gcc9 gcc10 gcc11 gcc12 gcc13`
- [ ] `release-small.nix` (3,412 jobs total)

I :heart: my new build farm

amjoseph-nixpkgs commented:
wineWowPackages.waylandFull.stdenv.cc.cc on x86_64-linux — [Success](https://github.com/NixOS/nixpkgs/pull/237856/checks?check_run_id=15000581773)

trofi commented:
Looks like https://github.com/NixOS/nixpkgs/pull/241980 was lost (and maybe more?)

amjoseph-nixpkgs commented:
> Looks like #241980 was lost

Fixed, thanks for noticing.

amjoseph-nixpkgs commented:
@ofborg build perlPackages.XMLLibXML

amjoseph-nixpkgs commented:
Rebased to fix merge conflict.

Artturin reviewed (approved):
Let's see if anything breaks

Built pkgsCross.aarch64-multiplatform.bash

Artturin commented on pkgs/development/compilers/gcc/common/builder.nix:
```diff
@@ -0,0 +1,289 @@
+            "CXXFLAGS_FOR_BUILD=$EXTRA_FLAGS_FOR_BUILD $EXTRA_LDFLAGS_FOR_BUILD"
+            "FLAGS_FOR_BUILD=$EXTRA_FLAGS_FOR_BUILD $EXTRA_LDFLAGS_FOR_BUILD"
+
+            # It seems there is a bug in GCC 5
+            #"CFLAGS=$EXTRA_FLAGS $EXTRA_LDFLAGS"
+            #"CXXFLAGS=$EXTRA_FLAGS $EXTRA_LDFLAGS"
```
This has been commented since it's introduction in 2017 https://github.com/nixos/nixpkgs/commit/d4595b38e90790a2b8ee01a7127889747ba05bb7#diff-7e60bf21d59f0d31c6d1a2c867692b9bcbad6dc8d8611e06e2e7bb91a9e07877R116-R118
